File: lib/f1_types/packet_0_car_motion_data.py
# MIT License
#
# Copyright (c) [2024] [Ashwin Natarajan]
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


import struct
from typing import Dict, Any
from .common import _split_list, PacketHeader, InvalidPacketLengthError, List

# --------------------- CLASS DEFINITIONS --------------------------------------

class CarMotionData:
    """
    A class for parsing the Car Motion Data of a telemetry packet in a racing game.
    The car motion data structure is as follows:

    Attributes:
        - m_worldPositionX (float): World space X position - metres
        - m_worldPositionY (float): World space Y position - metres
        - m_worldPositionZ (float): World space Z position - metres
        - m_worldVelocityX (float): Velocity in world space X - metres/s
        - m_worldVelocityY (float): Velocity in world space Y - metres/s
        - m_worldVelocityZ (float): Velocity in world space Z - metres/s
        - m_worldForwardDirX (int): World space forward X direction (normalised)
        - m_worldForwardDirY (int): World space forward Y direction (normalised)
        - m_worldForwardDirZ (int): World space forward X direction (normalised)
        - m_worldRightDirX (int): World space right X direction (normalised)
        - m_worldRightDirY (int): World space right Y direction (normalised)
        - m_worldRightDirZ (int): World space right Z direction (normalised)
        - m_gForceLateral (float): Lateral G-Force component
        - m_gForceLongitudinal (float): Longitudinal G-Force component
        - m_gForceVertical (float): Vertical G-Force component
        - m_yaw (float): Yaw angle in radians
        - m_pitch (float): Pitch angle in radians
        - m_roll (float): Roll angle in radians
    """

    PACKET_FORMAT = ("<"
        "f" # float - World space X position - metres
        "f" # float - World space Y position
        "f" # float - World space Z position
        "f" # float - Velocity in world space X – metres/s
        "f" # float - Velocity in world space Y
        "f" # float - Velocity in world space Z
        "h" # int16 - World space forward X direction (normalised)
        "h" # int16 - World space forward Y direction (normalised)
        "h" # int16 - World space forward Z direction (normalised)
        "h" # int16 - World space right X direction (normalised)
        "h" # int16 - World space right Y direction (normalised)
        "h" # int16 - World space right Z direction (normalised)
        "f" # float - Lateral G-Force component
        "f" # float - Longitudinal G-Force component
        "f" # float - Vertical G-Force component
        "f" # float - Yaw angle in radians
        "f" # float - Pitch angle in radians
        "f" # float - Roll angle in radians
    )
    PACKET_LEN: int = struct.calcsize(PACKET_FORMAT)

    def __init__(self, data: bytes) -> None:
        """A class for parsing the data related to the motion of the F1 car

        Args:
            data (List[bytes]): list containing the raw bytes for this packet
        """

        # Declare the data type hints
        self.m_worldPositionX: float
        self.m_worldPositionY: float
        self.m_worldPositionZ: float
        self.m_worldVelocityX: float
        self.m_worldVelocityY: float
        self.m_worldVelocityZ: float
        self.m_worldForwardDirX: int
        self.m_worldForwardDirY: int
        self.m_worldForwardDirZ: int
        self.m_worldRightDirX: int
        self.m_worldRightDirY: int
        self.m_worldRightDirZ: int
        self.m_gForceLateral: float
        self.m_gForceLongitudinal: float
        self.m_gForceVertical: float
        self.m_yaw: float
        self.m_pitch: float
        self.m_roll: float

        # Now, unpack the data and populate the members
        self.m_worldPositionX, self.m_worldPositionY, self.m_worldPositionZ, \
        self.m_worldVelocityX, self.m_worldVelocityY, self.m_worldVelocityZ, \
        self.m_worldForwardDirX, self.m_worldForwardDirY, self.m_worldForwardDirZ, \
        self.m_worldRightDirX, self.m_worldRightDirY, self.m_worldRightDirZ, \
        self.m_gForceLateral, self.m_gForceLongitudinal, self.m_gForceVertical, \
        self.m_yaw, self.m_pitch, self.m_roll = struct.unpack(self.PACKET_FORMAT, data)

    def __str__(self) -> str:
        """Return a formatted string representing the CarMotionData object

        Returns:
            str - string representation of this object
        """
        return (
            f"CarMotionData("
            f"World Position: ({self.m_worldPositionX}, {self.m_worldPositionY}, {self.m_worldPositionZ}), "
            f"World Velocity: ({self.m_worldVelocityX}, {self.m_worldVelocityY}, {self.m_worldVelocityZ}), "
            f"World Forward Dir: ({self.m_worldForwardDirX}, {self.m_worldForwardDirY}, {self.m_worldForwardDirZ}), "
            f"World Right Dir: ({self.m_worldRightDirX}, {self.m_worldRightDirY}, {self.m_worldRightDirZ}), "
            f"G-Force Lateral: {self.m_gForceLateral}, "
            f"G-Force Longitudinal: {self.m_gForceLongitudinal}, "
            f"G-Force Vertical: {self.m_gForceVertical}, "
            f"Yaw: {self.m_yaw}, "
            f"Pitch: {self.m_pitch}, "
            f"Roll: {self.m_roll})"
        )

    def toJSON(self) -> Dict[str, Any]:
        """Converts the CarMotionData object to a dictionary suitable for JSON serialization.

        Returns:
            Dict[str, Any]: A dictionary representing the JSON-compatible data.
        """
        return {
            "world-position": {
                "x": self.m_worldPositionX,
                "y": self.m_worldPositionY,
                "z": self.m_worldPositionZ
            },
            "world-velocity": {
                "x": self.m_worldVelocityX,
                "y": self.m_worldVelocityY,
                "z": self.m_worldVelocityZ
            },
            "world-forward-dir": {
                "x": self.m_worldForwardDirX,
                "y": self.m_worldForwardDirY,
                "z": self.m_worldForwardDirZ
            },
            "world-right-dir": {
                "x": self.m_worldRightDirX,
                "y": self.m_worldRightDirY,
                "z": self.m_worldRightDirZ
            },
            "g-force": {
                "lateral": self.m_gForceLateral,
                "longitudinal": self.m_gForceLongitudinal,
                "vertical": self.m_gForceVertical
            },
            "orientation": {
                "yaw": self.m_yaw,
                "pitch": self.m_pitch,
                "roll": self.m_roll
            }
        }

    def __eq__(self, other: Any) -> bool:
        """Check if this CarMotionData is equal to another.

        Args:
            other (Any): The object to compare against.

        Returns:
            bool: True if equal, False otherwise.
        """
        if not isinstance(other, CarMotionData):
            return NotImplemented
        return (
            self.m_worldPositionX == other.m_worldPositionX and
            self.m_worldPositionY == other.m_worldPositionY and
            self.m_worldPositionZ == other.m_worldPositionZ and
            self.m_worldVelocityX == other.m_worldVelocityX and
            self.m_worldVelocityY == other.m_worldVelocityY and
            self.m_worldVelocityZ == other.m_worldVelocityZ and
            self.m_worldForwardDirX == other.m_worldForwardDirX and
            self.m_worldForwardDirY == other.m_worldForwardDirY and
            self.m_worldForwardDirZ == other.m_worldForwardDirZ and
            self.m_worldRightDirX == other.m_worldRightDirX and
            self.m_worldRightDirY == other.m_worldRightDirY and
            self.m_worldRightDirZ == other.m_worldRightDirZ and
            self.m_gForceLateral == other.m_gForceLateral and
            self.m_gForceLongitudinal == other.m_gForceLongitudinal and
            self.m_gForceVertical == other.m_gForceVertical and
            self.m_yaw == other.m_yaw and
            self.m_pitch == other.m_pitch and
            self.m_roll == other.m_roll
        )

    def __ne__(self, other: Any) -> bool:
        """Check if this CarMotionData is not equal to another.

        Args:
            other (Any): The object to compare against.

        Returns:
            bool: True if not equal, False otherwise.
        """
        return not self.__eq__(other)

    def to_bytes(self) -> bytes:
        """Serialize the CarMotionData object to bytes based on PACKET_FORMAT.

        Returns:
            bytes: The serialized bytes.
        """
        return struct.pack(self.PACKET_FORMAT,
                           self.m_worldPositionX, self.m_worldPositionY, self.m_worldPositionZ,
                           self.m_worldVelocityX, self.m_worldVelocityY, self.m_worldVelocityZ,
                           self.m_worldForwardDirX, self.m_worldForwardDirY, self.m_worldForwardDirZ,
                           self.m_worldRightDirX, self.m_worldRightDirY, self.m_worldRightDirZ,
                           self.m_gForceLateral, self.m_gForceLongitudinal, self.m_gForceVertical,
                           self.m_yaw, self.m_pitch, self.m_roll)

    @classmethod
    def from_values(cls, world_position_x: float, world_position_y: float, world_position_z: float,
                    world_velocity_x: float, world_velocity_y: float, world_velocity_z: float,
                    world_forward_dir_x: int, world_forward_dir_y: int, world_forward_dir_z: int,
                    world_right_dir_x: int, world_right_dir_y: int, world_right_dir_z: int,
                    g_force_lateral: float, g_force_longitudinal: float, g_force_vertical: float,
                    yaw: float, pitch: float, roll: float) -> 'CarMotionData':
        """Create a CarMotionData object from individual values.

        Args:
            world_position_x (float): World space X position - metres
            world_position_y (float): World space Y position - metres
            world_position_z (float): World space Z position - metres
            world_velocity_x (float): Velocity in world space X - metres/s
            world_velocity_y (float): Velocity in world space Y - metres/s
            world_velocity_z (float): Velocity in world space Z - metres/s
            world_forward_dir_x (int): World space forward X direction (normalised)
            world_forward_dir_y (int): World space forward Y direction (normalised)
            world_forward_dir_z (int): World space forward X direction (normalised)
            world_right_dir_x (int): World space right X direction (normalised)
            world_right_dir_y (int): World space right Y direction (normalised)
            world_right_dir_z (int): World space right Z direction (normalised)
            g_force_lateral (float): Lateral G-Force component
            g_force_longitudinal (float): Longitudinal G-Force component
            g_force_vertical (float): Vertical G-Force component
            yaw (float): Yaw angle in radians
            pitch (float): Pitch angle in radians
            roll (float): Roll angle in radians

        Returns:
            CarMotionData: A CarMotionData object initialized with the provided values.
        """
        data = struct.pack(CarMotionData.PACKET_FORMAT,
                           world_position_x, world_position_y, world_position_z,
                           world_velocity_x, world_velocity_y, world_velocity_z,
                           world_forward_dir_x, world_forward_dir_y, world_forward_dir_z,
                           world_right_dir_x, world_right_dir_y, world_right_dir_z,
                           g_force_lateral, g_force_longitudinal, g_force_vertical,
                           yaw, pitch, roll)
        return cls(data)

class PacketMotionData:
    """A class for parsing the Motion Data Packet of a telemetry packet in a racing game.

    Args:
        header (PacketHeader): Incoming packet header
        packet (List[bytes]): list containing the raw bytes for this packet

    Attributes:
    - m_header (PacketHeader): The header of the telemetry packet.
    - m_car_motion_data (list): List of CarMotionData objects containing data for all cars on track.

    Raises:
        InvalidPacketLengthError: If received length is not as per expectation
    """

    def __init__(self, header:PacketHeader, packet: bytes) -> None:
        """Construct the PacketMotionData object from the given packet payload

        Args:
            header (PacketHeader): the parsed header object
            packet (bytes): The packet containing only the payload (header must be stripped)

        Raises:
            InvalidPacketLengthError: If number of bytes is not as per expectation
        """

        self.m_header: PacketHeader = header       # PacketHeader

        if ((len(packet) % CarMotionData.PACKET_LEN) != 0):
            raise InvalidPacketLengthError("Received packet length " + str(len(packet)) + " is not a multiple of " +
                                            str(CarMotionData.PACKET_LEN))
        self.m_carMotionData: List[CarMotionData] = []
        motion_data_packet_per_car = _split_list(packet, CarMotionData.PACKET_LEN)
        for motion_data_packet in motion_data_packet_per_car:
            self.m_carMotionData.append(CarMotionData(motion_data_packet))

    def __str__(self) -> str:
        """
        Return a string representation of the PacketMotionData instance.

        Returns:
        - str: String representation of PacketMotionData.
        """
        car_motion_data_str = ", ".join(str(car) for car in self.m_carMotionData)
        return f"PacketMotionData(Header: {str(self.m_header)}, CarMotionData: [{car_motion_data_str}])"

    def toJSON(self, include_header: bool=False) -> Dict[str, Any]:
        """Converts the PacketMotionData object to a dictionary suitable for JSON serialization.

        Arguments:
            - include_header - Whether the header dump must be included in the JSON

        Returns:
            Dict[str, Any]: A dictionary representing the JSON-compatible data.
        """

        json_data = {
            "car-motion-data": [car.toJSON() for car in self.m_carMotionData]
        }

        if include_header:
            json_data["header"] = self.m_header.toJSON()

        return json_data

    def __eq__(self, other: Any) -> bool:
        """Check if this PacketMotionData is equal to another.

        Args:
            other (Any): The object to compare against.

        Returns:
            bool: True if equal, False otherwise.
        """
        if not isinstance(other, PacketMotionData):
            return NotImplemented

        return self.m_header == other.m_header and self.m_carMotionData == other.m_carMotionData

    def __ne__(self, other: Any) -> bool:
        """Check if this PacketMotionData is not equal to another.

        Args:
            other (Any): The object to compare against.

        Returns:
            bool: True if not equal, False otherwise.
        """
        return not self.__eq__(other)

    def to_bytes(self) -> bytes:
        """Serialize the PacketMotionData object to bytes based on PACKET_FORMAT.

        Returns:
            bytes: The serialized bytes.
        """
        return self.m_header.to_bytes() + b''.join([car.to_bytes() for car in self.m_carMotionData])

    @classmethod
    def from_values(cls, header: PacketHeader, car_motion_data: List[CarMotionData]) -> 'PacketMotionData':
        """Create a PacketMotionData object from individual values.

        Args:
            header (PacketHeader): The header of the telemetry packet.
            car_motion_data (List[CarMotionData]): List of CarMotionData objects containing data for all cars on track.

        Returns:
            PacketMotionData: A PacketMotionData object initialized with the provided values.
        """
        return cls(header, b''.join([car.to_bytes() for car in car_motion_data]))

File: apps/backend/rust/crates/f1-types/src/packet_0_car_motion_data.rs
use crate::PacketHeader;
use crate::errors::InvalidPacketLengthError;
use serde::Serialize;
use serde::ser::{SerializeMap, Serializer};
use std::fmt;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct CarMotionData {
    pub world_position_x: f32,
    pub world_position_y: f32,
    pub world_position_z: f32,
    pub world_velocity_x: f32,
    pub world_velocity_y: f32,
    pub world_velocity_z: f32,
    pub world_forward_dir_x: i16,
    pub world_forward_dir_y: i16,
    pub world_forward_dir_z: i16,
    pub world_right_dir_x: i16,
    pub world_right_dir_y: i16,
    pub world_right_dir_z: i16,
    pub g_force_lateral: f32,
    pub g_force_longitudinal: f32,
    pub g_force_vertical: f32,
    pub yaw: f32,
    pub pitch: f32,
    pub roll: f32,
}

impl CarMotionData {
    pub const PACKET_LEN: usize = 60;

    #[allow(clippy::too_many_arguments)]
    pub fn from_values(
        world_position_x: f32,
        world_position_y: f32,
        world_position_z: f32,
        world_velocity_x: f32,
        world_velocity_y: f32,
        world_velocity_z: f32,
        world_forward_dir_x: i16,
        world_forward_dir_y: i16,
        world_forward_dir_z: i16,
        world_right_dir_x: i16,
        world_right_dir_y: i16,
        world_right_dir_z: i16,
        g_force_lateral: f32,
        g_force_longitudinal: f32,
        g_force_vertical: f32,
        yaw: f32,
        pitch: f32,
        roll: f32,
    ) -> Self {
        Self {
            world_position_x,
            world_position_y,
            world_position_z,
            world_velocity_x,
            world_velocity_y,
            world_velocity_z,
            world_forward_dir_x,
            world_forward_dir_y,
            world_forward_dir_z,
            world_right_dir_x,
            world_right_dir_y,
            world_right_dir_z,
            g_force_lateral,
            g_force_longitudinal,
            g_force_vertical,
            yaw,
            pitch,
            roll,
        }
    }

    pub fn parse(data: &[u8]) -> Result<Self, InvalidPacketLengthError> {
        if data.len() != Self::PACKET_LEN {
            return Err(InvalidPacketLengthError::new(format!(
                "Received packet length {} expected {}",
                data.len(),
                Self::PACKET_LEN
            )));
        }

        Ok(Self {
            world_position_x: f32::from_le_bytes(
                data[0..4].try_into().expect("fixed slice length"),
            ),
            world_position_y: f32::from_le_bytes(
                data[4..8].try_into().expect("fixed slice length"),
            ),
            world_position_z: f32::from_le_bytes(
                data[8..12].try_into().expect("fixed slice length"),
            ),
            world_velocity_x: f32::from_le_bytes(
                data[12..16].try_into().expect("fixed slice length"),
            ),
            world_velocity_y: f32::from_le_bytes(
                data[16..20].try_into().expect("fixed slice length"),
            ),
            world_velocity_z: f32::from_le_bytes(
                data[20..24].try_into().expect("fixed slice length"),
            ),
            world_forward_dir_x: i16::from_le_bytes(
                data[24..26].try_into().expect("fixed slice length"),
            ),
            world_forward_dir_y: i16::from_le_bytes(
                data[26..28].try_into().expect("fixed slice length"),
            ),
            world_forward_dir_z: i16::from_le_bytes(
                data[28..30].try_into().expect("fixed slice length"),
            ),
            world_right_dir_x: i16::from_le_bytes(
                data[30..32].try_into().expect("fixed slice length"),
            ),
            world_right_dir_y: i16::from_le_bytes(
                data[32..34].try_into().expect("fixed slice length"),
            ),
            world_right_dir_z: i16::from_le_bytes(
                data[34..36].try_into().expect("fixed slice length"),
            ),
            g_force_lateral: f32::from_le_bytes(
                data[36..40].try_into().expect("fixed slice length"),
            ),
            g_force_longitudinal: f32::from_le_bytes(
                data[40..44].try_into().expect("fixed slice length"),
            ),
            g_force_vertical: f32::from_le_bytes(
                data[44..48].try_into().expect("fixed slice length"),
            ),
            yaw: f32::from_le_bytes(data[48..52].try_into().expect("fixed slice length")),
            pitch: f32::from_le_bytes(data[52..56].try_into().expect("fixed slice length")),
            roll: f32::from_le_bytes(data[56..60].try_into().expect("fixed slice length")),
        })
    }

    pub fn to_bytes(&self) -> [u8; Self::PACKET_LEN] {
        let mut bytes = [0u8; Self::PACKET_LEN];
        bytes[0..4].copy_from_slice(&self.world_position_x.to_le_bytes());
        bytes[4..8].copy_from_slice(&self.world_position_y.to_le_bytes());
        bytes[8..12].copy_from_slice(&self.world_position_z.to_le_bytes());
        bytes[12..16].copy_from_slice(&self.world_velocity_x.to_le_bytes());
        bytes[16..20].copy_from_slice(&self.world_velocity_y.to_le_bytes());
        bytes[20..24].copy_from_slice(&self.world_velocity_z.to_le_bytes());
        bytes[24..26].copy_from_slice(&self.world_forward_dir_x.to_le_bytes());
        bytes[26..28].copy_from_slice(&self.world_forward_dir_y.to_le_bytes());
        bytes[28..30].copy_from_slice(&self.world_forward_dir_z.to_le_bytes());
        bytes[30..32].copy_from_slice(&self.world_right_dir_x.to_le_bytes());
        bytes[32..34].copy_from_slice(&self.world_right_dir_y.to_le_bytes());
        bytes[34..36].copy_from_slice(&self.world_right_dir_z.to_le_bytes());
        bytes[36..40].copy_from_slice(&self.g_force_lateral.to_le_bytes());
        bytes[40..44].copy_from_slice(&self.g_force_longitudinal.to_le_bytes());
        bytes[44..48].copy_from_slice(&self.g_force_vertical.to_le_bytes());
        bytes[48..52].copy_from_slice(&self.yaw.to_le_bytes());
        bytes[52..56].copy_from_slice(&self.pitch.to_le_bytes());
        bytes[56..60].copy_from_slice(&self.roll.to_le_bytes());
        bytes
    }
}

impl fmt::Display for CarMotionData {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "CarMotionData(World Position: ({}, {}, {}), World Velocity: ({}, {}, {}), World Forward Dir: ({}, {}, {}), World Right Dir: ({}, {}, {}), G-Force Lateral: {}, G-Force Longitudinal: {}, G-Force Vertical: {}, Yaw: {}, Pitch: {}, Roll: {})",
            self.world_position_x,
            self.world_position_y,
            self.world_position_z,
            self.world_velocity_x,
            self.world_velocity_y,
            self.world_velocity_z,
            self.world_forward_dir_x,
            self.world_forward_dir_y,
            self.world_forward_dir_z,
            self.world_right_dir_x,
            self.world_right_dir_y,
            self.world_right_dir_z,
            self.g_force_lateral,
            self.g_force_longitudinal,
            self.g_force_vertical,
            self.yaw,
            self.pitch,
            self.roll
        )
    }
}

impl Serialize for CarMotionData {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut map = serializer.serialize_map(Some(6))?;
        map.serialize_entry(
            "world-position",
            &Axis3f {
                x: self.world_position_x,
                y: self.world_position_y,
                z: self.world_position_z,
            },
        )?;
        map.serialize_entry(
            "world-velocity",
            &Axis3f {
                x: self.world_velocity_x,
                y: self.world_velocity_y,
                z: self.world_velocity_z,
            },
        )?;
        map.serialize_entry(
            "world-forward-dir",
            &Axis3i16 {
                x: self.world_forward_dir_x,
                y: self.world_forward_dir_y,
                z: self.world_forward_dir_z,
            },
        )?;
        map.serialize_entry(
            "world-right-dir",
            &Axis3i16 {
                x: self.world_right_dir_x,
                y: self.world_right_dir_y,
                z: self.world_right_dir_z,
            },
        )?;
        map.serialize_entry(
            "g-force",
            &GForce {
                lateral: self.g_force_lateral,
                longitudinal: self.g_force_longitudinal,
                vertical: self.g_force_vertical,
            },
        )?;
        map.serialize_entry(
            "orientation",
            &Orientation {
                yaw: self.yaw,
                pitch: self.pitch,
                roll: self.roll,
            },
        )?;
        map.end()
    }
}

#[derive(Serialize)]
struct Axis3f {
    x: f32,
    y: f32,
    z: f32,
}

#[derive(Serialize)]
struct Axis3i16 {
    x: i16,
    y: i16,
    z: i16,
}

#[derive(Serialize)]
struct GForce {
    lateral: f32,
    longitudinal: f32,
    vertical: f32,
}

#[derive(Serialize)]
struct Orientation {
    yaw: f32,
    pitch: f32,
    roll: f32,
}

#[derive(Clone, Debug, PartialEq)]
pub struct PacketMotionData {
    pub header: PacketHeader,
    pub car_motion_data: Vec<CarMotionData>,
}

impl PacketMotionData {
    pub const MAX_CARS: usize = 22;
    pub const PAYLOAD_LEN: usize = Self::MAX_CARS * CarMotionData::PACKET_LEN;

    pub fn parse(header: PacketHeader, packet: &[u8]) -> Result<Self, InvalidPacketLengthError> {
        if packet.len() != Self::PAYLOAD_LEN {
            return Err(InvalidPacketLengthError::new(format!(
                "Received packet length {} is not equal to expected {}",
                packet.len(),
                Self::PAYLOAD_LEN
            )));
        }

        let car_motion_data = packet
            .chunks_exact(CarMotionData::PACKET_LEN)
            .map(CarMotionData::parse)
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Self {
            header,
            car_motion_data,
        })
    }

    pub fn from_values(header: PacketHeader, car_motion_data: Vec<CarMotionData>) -> Self {
        Self {
            header,
            car_motion_data,
        }
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(PacketHeader::PACKET_LEN + Self::PAYLOAD_LEN);
        bytes.extend_from_slice(&self.header.to_bytes());
        for car in &self.car_motion_data {
            bytes.extend_from_slice(&car.to_bytes());
        }
        bytes
    }
}

impl fmt::Display for PacketMotionData {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let car_motion_data = self
            .car_motion_data
            .iter()
            .map(ToString::to_string)
            .collect::<Vec<_>>()
            .join(", ");
        write!(
            f,
            "PacketMotionData(Header: {}, CarMotionData: [{}])",
            self.header, car_motion_data
        )
    }
}

impl Serialize for PacketMotionData {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut map = serializer.serialize_map(Some(1))?;
        map.serialize_entry("car-motion-data", &self.car_motion_data)?;
        map.end()
    }
}
